(maint) Pass the appropriate Puppetfile path to module loader

mwaggett opened the pull request:
see individual commit messages for details

mwaggett committed:
(maint) Move Puppetfile logging

This commit moves the Puppetfile logging to the `R10K::ModuleLoader::Puppetfile`
since that is what does the actual loading.

mwaggett committed:
(maint) Rename `@puppetfile` to `@puppetfile_path` for clarity

justinstoller commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -41,11 +45,14 @@ def initialize(basedir:,
         @purge_exclusions = []
+
+        logger.info _("Using Puppetfile '%{puppetfile}'") % {puppetfile: @puppetfile_path}
+        logger.info _("Using moduledir '%{moduledir}'") % {moduledir: @moduledir}
```
I think this should be at debug.

Magisus replied:
I was thinking that too.

mwaggett replied:
Both or just the `moduledir` one? Either way is fine with me, just that the `puppetfile` one was at `info` before, which is why I put them both at that level here, but I don't disagree that debug makes sense.

justinstoller replied:
I'd just do the moduledir one, I think they both feel like debug statements but since the puppetfile one was info we should probably unfortunately keep it there.

mwaggett replied:
updated

justinstoller commented on lib/r10k/puppetfile.rb:
```diff
@@ -104,10 +104,10 @@ def load(default_branch_override = nil)
     if self.loaded?
       return @loaded_content
     else
-      if !File.readable?(@puppetfile_path)
```
Previous to this a missing Puppetfile would cause this portion the deployment to be a noop, while a syntax error in your Puppetfile would still cause the run to fail, right?

It seems we may want to either fix the `@puppetfile_path` to be the same kind of path that the loader will use. Or catch SystemCallError (or whatever else a missing/unreadable file might throw) here?

mwaggett replied:
yes.

I have another PR I'm gonna put up that will change the path used here, but it seemed better to me to do the validation/rescuing in the method that would actually throw, especially since the `ModuleLoader` methods are the ones we actually plan to keep. Are you just thinking that, by putting it here, we'd avoid unnecessarily trying to load altogether? Or is there another reason you think that stuff should go here?

justinstoller replied:
I think previously the run would fail if their were syntax errors in the Puppetfile, while now those errors will be printed at debug and the run will pass. (I might be wrong about that, but if that's the case we shouldn't do it).

Good point about the exception handling, I was just spitballing ideas that would keep us from doing a blanket rescue here.

mwaggett replied:
yeah, you're right. I guess I wasn't really clear on _why_ we just log and move on if the Puppetfile is unreadable but throw if there's another kind of error - I assumed that we didn't want to fail the run if there was anything wrong with the Puppetfile in whatever case(s) `load` is used (as opposed to `load!`), but didn't investigate at all. I can bring back the old behavior if you think that's better. I guess a change would probably be considered breaking by _someone_ huh..

justinstoller replied:
So I don't think you need a Puppetfile in an environment if you either don't have dependencies or you vendor them all. In that case the load becomes a no-op and the run continues. However, the attempting to deploy external dependencies is an opt-in feature so maybe it should be an error if you opt-in to deploying modules but there is not Puppetfile? I suppose you could do a deploy all kinda thing and want it to deploy modules if the Puppetfile exists in an environment?

mwaggett replied:
I removed this change for now.

mwaggett committed:
(maint) Pass the appropriate Puppetfile path to module loader

Previously, we always passed the `puppetfile_name` from the `R10K::Puppetfile`
class to `R10K::ModuleLoader::Puppetfile`. That was an attempt to fix a bug
where relative Puppetfile paths were not getting resolved correctly by the
module loader (https://github.com/puppetlabs/r10k/commit/9082511b8e862e3f9189470cc19986dd61576206).
That, however, unearthed another bug: when instantiating a new instance of
`R10K::Puppetfile` with the `puppetfile_path` option, that path would get
completely ignored in favor of the default `puppetfile_name` value. This broke
usage of the `r10k puppetfile purge` command.

This commit ensures that we pass the appropriate Puppetfile path to the module
loader - we pass the `puppetfile_path` if it exists, otherwise we pass the
`puppetfile_name`. In order to maintain the fix for the original bug around
relative Puppetfile paths, we no longer attempt to resolve the `puppetfile_path`
in the `R10K::Puppetfile` class, since `R10K::ModuleLoader::Puppetfile` already
handles that resolution.

Magisus reviewed (approved):
I think this looks alright, but Justin has more context.

mwaggett committed:
(maint) Use production Puppetfile in `purge` acceptance test

Previously, we were using the remote version of the Puppetfile to test the
`r10k puppetfile purge` command. This commit updates the test to use the
deployed version of the Puppetfile, which is what a user would actually use.

justinstoller commented on spec/unit/puppetfile_spec.rb:
```diff
@@ -10,14 +10,35 @@
 
+describe R10K::Puppetfile do
+
+  relative_subject = described_class.new('/some/nonexistent/basedir',
```
I think these subjects should go in the `it` blocks, right?

mwaggett replied:
I copied this from the test above, which has the `subject` outside of the `it` block. Let me know if you still think I should move them.

justinstoller replied:
The above tests use the `subject` RSpec DSL method, see: https://relishapp.com/rspec/rspec-core/docs/subject/explicit-subject .

These appear to be local variables that for some reason are still be in scope when the it block is evaluated.

mwaggett replied:
yes, I know. I assume you're saying "yes, change it"?

mwaggett replied:
updated

justinstoller replied:
Yes.

justinstoller commented on spec/unit/puppetfile_spec.rb:
```diff
@@ -10,14 +10,35 @@
+describe R10K::Puppetfile do
+
+  describe "a custom relative puppetfile_path" do
+    relative_subject = described_class.new('/some/nonexistent/basedir',
```
These should be in the `it` block.
